crates/language-server/src/functionality/handlers.rs

```diff
@@ -0,0 +1,137 @@
+use crate::backend::Backend;
+
+use crate::backend::workspace::SyncableIngotFileContext;
+
+use common::InputDb;
+use futures::TryFutureExt;
+use fxhash::FxHashSet;
+
+use lsp_types::TextDocumentItem;
+use salsa::ParallelDatabase;
+
+use super::{capabilities::server_capabilities, hover::hover_helper};
+
+use crate::backend::workspace::{IngotFileContext, SyncableInputFile};
+
+use tracing::{error, info};
+
+impl Backend {
+    pub(super) async fn handle_initialized(
+        &mut self,
+        params: lsp_types::InitializeParams,
+        responder: tokio::sync::oneshot::Sender<
+            Result<lsp_types::InitializeResult, tower_lsp::jsonrpc::Error>,
+        >,
+    ) {
+        info!("initializing language server!");
+
+        let root = params.root_uri.unwrap().to_file_path().ok().unwrap();
+
+        let _ = self.workspace.set_workspace_root(&mut self.db, &root);
+        let _ = self.workspace.load_std_lib(&mut self.db, &root);
+        let _ = self.workspace.sync(&mut self.db);
+
+        let capabilities = server_capabilities();
+        let initialize_result = lsp_types::InitializeResult {
+            capabilities,
+            server_info: Some(lsp_types::ServerInfo {
+                name: String::from("fe-language-server"),
+                version: Some(String::from(env!("CARGO_PKG_VERSION"))),
+            }),
+        };
+        let _ = responder.send(Ok(initialize_result));
+    }
+
+    pub(super) async fn handle_shutdown(
+        &mut self,
+        responder: tokio::sync::oneshot::Sender<Result<(), tower_lsp::jsonrpc::Error>>,
+    ) {
+        info!("shutting down language server");
+        let _ = responder.send(Ok(()));
+    }
+
+    pub(super) async fn handle_deleted(
+        &mut self,
+        params: lsp_types::FileEvent,
+        tx_needs_diagnostics: tokio::sync::mpsc::UnboundedSender<String>,
+    ) {
+        let path = params.uri.to_file_path().unwrap();
+        info!("file deleted: {:?}", path);
+        let path = path.to_str().unwrap();
+        let _ = self
+            .workspace
+            .remove_input_for_file_path(&mut self.db, path);
+        let _ = tx_needs_diagnostics.send(path.to_string());
+    }
+
+    pub(super) async fn handle_change(
+        &mut self,
+        doc: TextDocumentItem,
+        tx_needs_diagnostics: tokio::sync::mpsc::UnboundedSender<String>,
+    ) {
+        info!("change detected: {:?}", doc.uri);
+        let path_buf = doc.uri.to_file_path().unwrap();
+        let path = path_buf.to_str().unwrap();
+        let contents = Some(doc.text);
+        if let Some(contents) = contents {
+            let input = self
+                .workspace
+                .touch_input_for_file_path(&mut self.db, path)
+                .unwrap();
+            let _ = input.sync_from_text(&mut self.db, contents);
+        }
+        let _ = tx_needs_diagnostics.send(path.to_string());
+    }
```

Comment:
(Associated change for the removed of the stream forking)

```suggestion
    pub(super) async fn handle_change(
        &mut self,
        change: FileChange,
        tx_needs_diagnostics: tokio::sync::mpsc::UnboundedSender<String>,
    ) {
        let path = change.uri.to_string();

        match change.kind {
            ChangeKind::Open(contents) => {
                info!("file opened: {:?}", &path);
                self.update_input_file_text(&path, contents);
            }
            ChangeKind::Create => {
                info!("file created: {:?}", &path);
                let contents = tokio::fs::read_to_string(&path).await.unwrap();
                self.update_input_file_text(&path, contents)
            }
            ChangeKind::Edit(contents) => {
                info!("file edited: {:?}", &path);
                let contents = if let Some(text) = contents {
                    text
                } else {
                    tokio::fs::read_to_string(&path).await.unwrap()
                };
                self.update_input_file_text(&path, contents);
            }
            ChangeKind::Delete => {
                info!("file deleted: {:?}", path);
                self.workspace
                    .remove_input_for_file_path(&mut self.db, &path)
                    .unwrap();
            }
        }
        tx_needs_diagnostics.send(path).unwrap();
    }
    
    fn update_input_file_text(&mut self, path: &str, contents: String) {
        let input = self
            .workspace
            .touch_input_for_file_path(&mut self.db, path)
            .unwrap();
        let _ = input.sync_from_text(&mut self.db, contents);
    }

```

Reply:
I used `tokio::fs::read_to_string` here, but that might introduce the possibility of changes being handled out of order.

I share Yoshi's skepticism of using async because of these tricky situations, but I remain open minded. If we did drop async, I think the "streamy" essence of the code could remain, we'd just eg explicitly spawn a thread (or "actor") that sequentially handles all change events that arrive on a channel.

Reply:
I haven't tried it yet but this looks great.  Can you help me understand how these changes might get handled out of order?  My thought is that by merging all those change events into a single stream we will enforce that they're processed serially in the order they were merged, at least relative to one another.

Reply:
(By the way, I shared my complex thoughts on the async stuff here: https://github.com/ethereum/fe/pull/979#discussion_r1545977502)